services/weatherApi.service.go

```diff
@@ -0,0 +1,174 @@
+package services
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/gofiber/fiber/v2"
+	"gorm.io/gorm"
+	"net/http"
+	"project_weather/controllers"
+	"project_weather/generated/dao/model"
+	"project_weather/repositories"
+	"project_weather/resources/dto"
+	"time"
+)
+
+const (
+	forecastUrl         = "http://api.weatherapi.com/v1/forecast.json?key=%s&q=%s&days=%v&aqi=no&alerts=no"
+	currentTimeUrl      = "http://api.weatherapi.com/v1/current.json?key=%s&q=%s&aqi=no"
+	templateDate        = "2006-01-02"
+	templateDateAndTime = "2006-01-02 15:04"
+)
+
+type WeatherApiService interface {
+	GetRoutes() []controllers.Route
+	GetWeatherByCity(ctx *fiber.Ctx) error
+	fetchWeatherData(url string) (*dto.WeatherDTO, error)
+	formatWeatherData(weatherData *dto.WeatherDTO, period string, cityID string) (interface{}, error)
+}
+
+var periodToDays = map[string]int{
+	"daily":   1,
+	"weekly":  7,
+	"monthly": 30,
+	"current": 0,
+}
+
+type weatherApiService struct {
+	cityRepo     repositories.CityDB
+	forecastRepo repositories.ForecastDB
+	client       *http.Client
+	apiKey       string
+	db           *gorm.DB
+}
+
+func NewWeatherService(client *http.Client, apiKey string, db *gorm.DB, cityDB repositories.CityDB, forecastDB repositories.ForecastDB) WeatherApiService {
+	return &weatherApiService{
+		client:       client,
+		apiKey:       apiKey,
+		db:           db,
+		cityRepo:     cityDB,
+		forecastRepo: forecastDB,
+	}
+}
+
+func (self *weatherApiService) GetRoutes() []controllers.Route {
+	return []controllers.Route{
+		{
+			Method:  http.MethodGet,
+			Path:    "/weather/:city/:period",
```

Comment:
What's the purpose of this endpoint? As far as I can see it fetches data from `api.weatherapi.com/v1`, registers a city if there is none and returns the data. How do you plan to populate the forecast resource?